@here/harp-mapview/lib/Utils.ts

```diff
@@ -242,53 +243,182 @@ export namespace MapViewUtils {
+        // two rays C->R and C->M, i.e. the angle RCM. This angle remains constant when tilting with
+        // a fixed `offsetX` and `offsetY`. It is calculated by using the intersection of the two
+        // ways with the earth. Where:
+        // MRC is 90 + `angleBetweenNormals` - tilt at R.
+        // CMR is 90 + `tilt`
+        // RCM is 180 - CRM - RMC, aka `angleBetweenRotationTargetAndMapTarget`
+        // Note the use of `angleBetweenNormals` to ensure this works for spherical projections.
+        const angleBetweenRotationTargetAndMapTarget =
+            ninetyRad * 2 -
+            (ninetyRad + angleBetweenNormals - rotationCenterTilt + ninetyRad + tilt);
+
+        // This uses the same formula as above, just we change `tilt` to the maxTiltAngle,
+        // i.e we want to find the greatest angle at the rotation target that gives us the max
+        // angle at the map center target.
+        // i.e. CMR is 90 + `maxTiltAngle`
+        const maxRotationTargetAngle =
+            ninetyRad * 2 - angleBetweenRotationTargetAndMapTarget - ninetyRad - maxTiltAngle;
+
+        // Converting the `maxRotationTargetAngle` back to a tilt is as easy as subtracting it from
+        // 90 and the `angleBetweenNormals`, i.e. this gives us the maximum allowed tilt at R
+        // that satisfies the `maxTiltAngle` constraint.
+        const maxTilt = ninetyRad + angleBetweenNormals - maxRotationTargetAngle;
+
+        const clampedDeltaTiltMapTarget =
+            MathUtils.clamp(deltaTilt + rotationCenterTilt, 0, maxTilt) - rotationCenterTilt;
+
+        const maxRotationTiltAngle = THREE.MathUtils.degToRad(89);
+        // Here we clamp to 89 degrees, to prevent it intersecting with the world at 90. This is
+        // possible for example when the R position is near the horizon. If the angle RCM is say 5
+        // degrees, then an angle of say 89 degrees at R, plus 5 degrees means the tilt at M would
+        // be 84 degrees. If the `maxTiltAngle` is however 89, then we still have what seems like 5
+        // degrees at M, so, without this check, we would go to 90 degrees.
+        const clampedDeltaTiltRotationTarget =
+            MathUtils.clamp(deltaTilt + rotationCenterTilt, 0, maxRotationTiltAngle) -
+            rotationCenterTilt;
+
+        // Math.min doesn't work, because we need to keep the sign of the delta as well.
+        const clampedDeltaTilt = Math.min(
+            clampedDeltaTiltMapTarget,
+            clampedDeltaTiltRotationTarget
+        );
+
+        // It can happen in some cases that the maxRotationTiltAngle is exceeded because the
+        // `clampedDeltaTilt` is taken from the smaller value clampedDeltaTiltMapTarget. We check
+        // that we don't exceed 0, because otherwise it will snap to 89, so we let the user ease
+        // back to 89, but not any further.
+        if (rotationCenterTilt + clampedDeltaTilt >= maxRotationTiltAngle) {
```

Comment:
Continuing from my last comment, this if condition would never be true:

rotationCenterTilt + clampedDeltaTilt =
rotationCenterTilt + clamp(deltaTilt + rotCenterTilt, 0, min(maxTilt,maxRotTiltAngle)) - rotCenterTilt =
clamp(deltaTilt + rotCenterTilt, 0, min(maxTilt,maxRotTiltAngle)) <= maxRotTiltAngle

Reply:
Yes, good point, I need to however still detect this condition when `rotationCenterTilt > maxRotationTiltAngle`, because then I want to return 0, otherwise the `clampedDeltaTilt` is negative and it will snap back to the 89 degrees at the rotation center world, which is a bit unnatural. Here, we should let the user not increase the `rotationCenterTilt`, but not snap it back either, I will add a condition above to prevent this case.

Reply:
Ok, but if rotationCenterTilt > maxRotationTiltAngle, it means you can return 0 deltaTilt much sooner in the function right?

Reply:
Yeah, check my latest commit, it should be in there.